### ISSUE TITLE
Inputs get dropped or ignored completely under certain conditions

### ISSUE THREAD

**Issue by Stoppedpuma** [woelper/oculante#374]
**Describe the bug**
Inputs seem to be dropped if they are pressed to fast and are ignored completely if oculante is moved to another workspace.

**To Reproduce**
Steps to reproduce the dropped behavior:
1. Open a directory containing multiple images
2. Switch between them quickly
3. Reproduced

Steps to reproduce the ignored behaviour:
1-2. Open a directory containing multiple images
2-2. Move oculante to another workspace on the same display
3-2. Reproduced

**Expected behavior**
Inputs to not be dropped.

**Screenshots / sample files**

There's a key overlay in the bottom right which can be used in frame-by-frame analysis if needed, otherwise it can still be noticed without analysing by frame.

https://github.com/woelper/oculante/assets/58333920/2d9d367d-d535-4021-9ae7-7d583dc70f90

https://github.com/woelper/oculante/assets/58333920/8338817c-c8a7-4656-857f-9b510b75d798

**Desktop (please complete the following information):**
 - OS: Arch Linux, Hyprland (Wayland compositor)

**Comment by woelper** [comment]
Maybe this is related to a design decision. In the past, Oculante blocked any loading operation until the image was loaded. For example if you pressed the button to load the next image, it could not be pressed / did not work until the image was loaded. 

This felt more "solid", but also was not feeling responsive when loading very large images, and it also was a problem when the images did not load at all. 

After that I decided that you could advance to the next image at any time. Since all the code is async, this would not block at all, and you can super quickly skip over a bunch of files, no matter how long they load.


The other issue with not registering commands at all might be related to the app losing input when moving it to another workspace. I suspect this is related to `winit` or some other dependency beyond my control, as Oculante does not handle that part directly but instead relies on a library to do that.

**Comment by Stoppedpuma** [comment]
I don't think this is related to the image being loaded, sometimes it just requires pressing the key twice for some reason I've found. I've also noticed this seems to happen a lot when "zen mode" is enabled and can result in the input being ignored completely as well every once in a while.

Regarding the input blocking problem though I've found out that this is caused by oculante not letting go of shift, pressing shift after moving it to another workspace fixes the input problem.

> Since all the code is async, this would not block at all, and you can super quickly skip over a bunch of files, no matter how long they load.

I noticed this and find it very useful, thanks for this!

**Comment by woelper** [comment]
That is very useful info, I will dig deeper on this!

Shortcuts are also provided by a library I use, so I might be limited. But I remember I have some special treatment for "modifier keys" such as Shift, maybe there is a bug.

**Comment by Stoppedpuma** [comment]
Another place where input gets ignored is while holding another key.

**Comment by Stoppedpuma** [comment]
This also might be a draw issue as the next / previous image will update after pressing another key. There's another weird behaviour with this as well with previous inputs that were dropped being handled again when the key that was dropped is pressed again after a while, in my case it's next / previous inputs sometimes being duplicated which I'm assuming is because of that.

**Comment by Stoppedpuma** [comment]
Keys do seem to be registered according to the debug log, but they aren't reporting doing anything. For example the log usually says:

DEBUG oculante] key down
oculante::shortcuts] Number of keys pressed: 1
oculante::shortcuts] Matched EditMode / "E"

When this issue happens it only says key down.

I believe this confirms that the reason this happens is keys being stuck upon leaving the window.

Regarding the image not updating, this is solved with enabling the "redraw every frame" option.

**Comment by lefp** [comment]
From a brief skim through the code, it seems like the Shift key getting stuck is probably a Notan bug, right?

**Comment by lefp** [comment]
This is a winit bug on Wayland.
https://github.com/rust-windowing/winit/issues/1272

**Comment by woelper** [comment]
Wow, thanks for investigating!

**Comment by lefp** [comment]
... Although I don't know why tapping shift after regaining focus doesn't clear the shift state. I would expect the shift key to get added and then removed from the HashSet they're using when that happens.
